### DISCUSSION TITLE
How to get user level access?

### DISCUSSION THREAD

**Discussion by YNedderhoff** [mastodon/mastodon#23377] [Q&A] [answered]
What is the proposed flow for getting long lived authorization with user-level access in a headless environment? 

I know I can manually create an authorization code using [`GET /oauth/authorize`](https://docs.joinmastodon.org/methods/oauth/#authorize) and then use that in [`POST /oauth/token`](https://docs.joinmastodon.org/methods/oauth/#token) with `grant_type=authorization_code`. However, this authorization code can expire. There seems to be no documentation in the about the TTL, or how to refresh it. 

Can someone give me some pointers?

**Comment by trwnh** [comment, marked as answer]
the authorization code is only used once, so it doesn't need to be refreshed. you exchange it for a user token and then the authorization code is invalidated.

**Reply by YNedderhoff** [reply]
And the user token does not expire?

**Reply by YNedderhoff** [reply]
Maybe just a quick summary of what I understand you're saying:

- get an auth code with `GET /oauth/authorize`
- get an access token with `POST /oauth/token`
- use that token forever?

So my question is: does it not expire, and if it does, how do I have the application refresh it without manual intervention required?

**Reply by trwnh** [reply]
yes, mastodon does not currently expire tokens. however, they can be revoked. pleroma's mastodon api implementation *does* expire tokens, i think? they support refresh tokens like so: https://docs-develop.pleroma.social/backend/development/API/differences_in_mastoapi_responses/#refreshing-a-token

**Reply by YNedderhoff** [reply]
I see - I guess that's the part that I didn't expect. I am not using pleroma so that's fine then. I suppose the documentation could be more explicit about that, given that I think the OAuth standard recommends short-lived access tokens and long-lived refresh tokens, plus an `expires_in` field?

Thanks though! I will try this.
